src/query/storage/index.go

```diff
@@ -136,26 +161,38 @@ func queryKey(m models.Matchers) []byte {
 	return key
 }
 
-// FetchQueryToM3Query converts an m3coordinator fetch query to an M3 query
-func FetchQueryToM3Query(fetchQuery *FetchQuery, cache *QueryConversionCache) (index.Query, error) {
+// FetchQueryToM3Query converts an m3coordinator fetch query to an M3 query.
+func FetchQueryToM3Query(
+	fetchQuery *FetchQuery,
+	cache *QueryConversionCache,
+) (index.Query, error) {
 	matchers := fetchQuery.TagMatchers
-	k := queryKey(matchers)
+	// If no matchers provided, explicitly set this to an AllQuery
+	if len(matchers) == 0 {
+		return index.Query{
+			Query: idx.NewAllQuery(),
+		}, nil
+	}
 
-	cache.mu.Lock()
-	defer cache.mu.Unlock()
+	k := queryKey(matchers)
+	cache.RLock()
 
 	if val, ok := cache.get(k); ok {
+		cache.RUnlock()
 		return index.Query{Query: val}, nil
 	}
 
+	cache.RUnlock()
```

Comment:
Can you simplify the above to not have to do two RUnlocks?

```go
k := queryKey(matchers)
cache.RLock()
val, ok := cache.get(k)
cache.RUnlock()

if ok { 
  return index.Query{Query: val}, nil
}

//....
```

Reply:
Sure, looks better